Jetpack Settings: Add Theme Enhancements card under Writing Settings for Jetpack sites

tyxla opened the pull request:
### Introduction

This PR is part of #9171. It introduces a Theme Enhancements card under Writing Settings. This card will be used to manage the settings of the Infinite Scroll and Mobile Theme Jetpack modules.

### Preview

#### Both modules activated, some settings are disabled
![](https://cldup.com/xCIIPi8iRc.png)

#### Initial loading
![](https://cldup.com/knDrQ3NSGK.png)

#### Loading while saving
![](https://cldup.com/vS2p_mDNWX.png)

#### Disabled modules (settings are hidden)
![](https://cldup.com/lBqsNkjADP.png)

### Testing

* Checkout this branch.
* Select one of your Jetpack sites with Jetpack 4.5 or newer.
* Go to `/settings/writing/$site`, where `$site` is the slug of your Jetpack site.
* Verify the Theme Enhancements card is displayed properly.
* Play with the main toggles that enable the modules, and verify they save correctly in your Jetpack site (note that notices are implemented separately in #10583).
* Verify that the settings of a module are displayed only when the module is active.
* Play around with the settings of each module and click "Save Settings". Verify they save properly in your Jetpack site.
* Verify the links of the info popovers are working as expected and lead to the right place.
* Test with a Jetpack site that has Jetpack 4.4.2 or older, and verify the card is not shown.
* Test with a WordPress.com site and verify that the card is not shown.

beaulebens commented:
Can we get a screenshot/mock of what we should be seeing for this?

beaulebens commented:
![screen shot 2016-12-22 at 2 29 15 pm](https://cloud.githubusercontent.com/assets/108942/21441000/1de5518e-c853-11e6-81f2-d90b391ad2d9.png)

You might not have done this one yet, but this appears to be the wrong helper/information link for this setting.

tyxla commented:
> Can we get a screenshot/mock of what we should be seeing for this?

Sure thing, I've updated the PR description to contain a screenshot. I'll also update with some testing instructions and additional information once the PR is ready to review.

> You might not have done this one yet, but this appears to be the wrong helper/information link for this setting.

Yeah, this was in progress, but I've now updated it, too. Could have missed it, so good catch!   :+1:

tyxla commented:
This PR works and saves data properly; however it will be best if we adapt it to using the bulk settings endpoint; this is being implemented in #10123.

tyxla commented:
Design-wise, this one is ready to review, however code-wise we have where to improve, specifically:
* A `wrapSettingsForm` HoC was introduced in #9899 that can be useful to reduce some boilerplate here.
* Reduxification of the Writing form is in process in #10691, it would be much better if we can take advantage of it in this PR.

MichaelArestad commented:
Looks good, but I would remove the external link icon on the tooltips.

Since these are all toggles, does there need to be a save button or can they just instantly work? (with proper success messages, etc)

tyxla commented:
Thank you for the feedback, @MichaelArestad!

> Looks good, but I would remove the external link icon on the tooltips.

Sounds good to me, I've removed the icon.

> Since these are all toggles, does there need to be a save button or can they just instantly work? (with proper success messages, etc)

Of course, we can make them work instantly. However if we do this, we'll have to do it for all site settings, not just for this card, because we'll want to keep a consistent user interface, and codebase. And this will involve a lot of personalized notices that currently don't exist. This can definitely be left for a future iteration IMHO.

a8ci18n commented on client/my-sites/site-settings/theme-enhancements/index.jsx:
```diff
@@ -0,0 +1,288 @@
+						<div className="theme-enhancements__module-settings is-indented">
+							{
+								this.renderToggle( 'wp_mobile_excerpt', translate(
+									'Use excerpts instead of full posts on front page and archive pages'
```
Alternate string suggestion: 
* `translate( 'Show excerpts on front page and on archive pages instead of full posts' )`&mdash; translations: **17**. 

[PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/jetpack-setings-theme-enhancements-card)

a8ci18n commented on client/my-sites/site-settings/theme-enhancements/index.jsx:
```diff
@@ -0,0 +1,288 @@
+				<JetpackModuleToggle
+					siteId={ selectedSiteId }
+					moduleSlug="minileven"
+					label={ translate( 'Optimize your site with a phone-friendly theme.' ) }
```
Alternate string suggestion: 
* `translate( 'Optimize your site with a mobile-friendly theme for tablets and phones.' )`&mdash; translations: **10**. 

[PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/jetpack-setings-theme-enhancements-card)

a8ci18n commented on client/my-sites/site-settings/theme-enhancements/index.jsx:
```diff
@@ -0,0 +1,288 @@
+				<div className="theme-enhancements__info-link-container">
+					<InfoPopover position={ 'left' }>
+						<ExternalLink href={ 'https://jetpack.com/support/mobile-theme' } target="_blank">
+							{ translate( 'Learn more about Mobile Theme' ) }
```
Alternate string suggestion: 
* `translate( 'Learn more about this {{a}}awesome theme{{/a}}.' )`&mdash; translations: **21**. 

[PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/jetpack-setings-theme-enhancements-card)

tyxla commented:
We're almost done here. This is waiting for #10691 to be merged.

MichaelArestad commented:
> This can definitely be left for a future iteration IMHO.

@tyxla Cool. Once we get these in place, it should be a high priority.

tyxla commented:
#10691 has been merged. This is ready for a code review; design review has already been done (thanks @MichaelArestad).
